x-pack/plugin/core/src/main/java/org/elasticsearch/license/StartBasicClusterTask.java

```diff
@@ -114,10 +121,29 @@ private License generateBasicLicense(ClusterState currentState) {
             .type(License.LicenseType.BASIC)
             .expiryDate(LicenseService.BASIC_SELF_GENERATED_LICENSE_EXPIRATION_MILLIS);
 
-        return SelfGeneratedLicense.create(specBuilder, currentState.nodes());
+        return SelfGeneratedLicense.create(specBuilder, discoveryNodes);
     }
 
     public String getDescription() {
         return description;
     }
+
+    static class Executor implements ClusterStateTaskExecutor<StartBasicClusterTask> {
+        @Override
+        public ClusterState execute(ClusterState currentState, List<TaskContext<StartBasicClusterTask>> taskContexts) throws Exception {
+            XPackPlugin.checkReadyForXPackCustomMetadata(currentState);
+            final LicensesMetadata originalLicensesMetadata = currentState.metadata().custom(LicensesMetadata.TYPE);
+            var currentLicensesMetadata = originalLicensesMetadata;
+            for (final var taskContext : taskContexts) {
+                currentLicensesMetadata = taskContext.getTask().execute(currentLicensesMetadata, currentState.nodes(), taskContext);
+            }
```

Comment:
> Is this loop exactly what the batching is for?

Yes this loop is in case there's a whole batch of tasks to complete. Admittedly unlikely but the function signature forces us to handle that possibility like this.

> What if the loop somehow crashes before the final clusterState is computed and updated?

The client's listeners aren't actually completed until much later. If this method throws an exception for some reason then all the listeners see the exception and the cluster state remains unchanged:

https://github.com/elastic/elasticsearch/blob/f3ec22a4d63f5af3c6f0ce0001952455226fffc9/server/src/main/java/org/elasticsearch/cluster/service/MasterService.java#L1010-L1029

Reply:
Thanks for the pointer, David. Now I see how the original listener is wrapped by `taskContext.success` which does not complete until its `onPublishXxx` or `onClusterStateXxx` method is called later on.

By briefly reading the code, it seems that `TaskContext.onBatchFailure` nullify the `publishListner`. Does not mean the listener basically just hangs till timeout happens (and client will see the timeout)?

I have another question which is kinda on the opposite side of my original one: There are individual responses that do not rely on the final success of the batched tasks. For example, if the request needs acknowledgement, technically this response does not need wait for the batch to complete. With batching, it is theorectically possible that this response gets cancelled because of other failures. I don't think it's a big issue, particularlly not for updating license. But I'd appreciate your thoughts on this. Thanks!

Reply:
> Does not mean the listener basically just hangs till timeout happens (and client will see the timeout)?

This listener is not completed indeed, instead we call the task's `onFailure` method as if the task was never executed, and that should also send a response back to the client.

> With batching, it is theorectically possible that this response gets cancelled because of other failures.

Yes that's true, you do have to be careful to distinguish failing a single task from failing the entire batch. Typically you want the former, but sometimes failing the batch might be the only possibility.